doc/developer/command_and_response_binary_encoding.md

```diff
@@ -0,0 +1,574 @@
+4. Complex types that are not defined by us (such as `DateTime`).
+
+The problem of encoding `$T` in a Protobuf-based binary format thereby decomposes into the problem of encoding instance of each of the above four classes.
+The following rules apply in general:
+
+- Types from (1) are trivially represented by their existing Protobuf counterpart.
+- Types from (3) are the primary focus of this guide.
+- Types from (2) only care about implementation steps (3) and (4), and types that belong to (4) don't care about step (1).
+- Types from (2) and (4) require a slightly different implementation mechanism for steps (3) and (4) compared to types from (3).
+  The [`chrono` types PR](https://github.com/MaterializeInc/materialize/pull/11801) is a good example of the latter.
+
+## Defining a `Protobuf` message for `Proto$T`
+
+This step is only needed if `$T` needs a custom Protobuf representation ([classes (3) and (4)](#type-classes)).
+The initial message definition of `Proto$T` can be derived schematically from the schape of `$T` (see [Appendix A](#appendix-a-deriving-an-initial-protobuf-message-for-t) for details).
+Here are the example contents of `expr/src/foo/bar.proto` for the running examples from [the previous section](#defining-a-rust-typet).
+
+```protobuf
+syntax = "proto3";
+
+import "repr/src/adt/char.proto";
+import "repr/src/chrono.proto";
+
+package mz_expr.foo.bar;
+
+// `$T` is a struct
+message ProtoMyStruct {
+    uint64 field_1 = 1;
+    uint64 field_2 = 2;
+    mz_repr.adt.char.ProtoCharLength field_3 = 3;
+    mz_repr.chrono.ProtoNaiveDate field_4 = 4;
+}
+
+// `$T` is an enum
+message ProtoMyEnum {
+    oneof kind {
+        uint64 var1 = 1;
+        uint64 var2 = 2;
+        mz_repr.adt.char.ProtoCharLength var3 = 3;
+        mz_repr.chrono.ProtoNaiveDate var4 = 4;
+    }
+}
+```
+
+## Extending `build.rs`
+
+This step is only needed if `$T` needs a custom Protobuf representation ([classes (3) and (4)](#type-classes)).
+
+```rust
+fn main() {
+    prost_build::Config::new()
+        // list paths to external types used in the compiled files
+        .extern_path(".mz_repr.adt.char", "::mz_repr::adt::char")
+        .extern_path(".mz_repr.chrono", "::mz_repr::chrono")
+        // snip (...)
+        // make the docstring linter happy
+        .type_attribute(".", "#[allow(missing_docs)]")
+        // list paths to `*.proto` files to be compiled
+        .compile_protos(
+            &[
+                "expr/src/foo/bar.proto",
+                // snip (...)
+            ],
+            &[".."],
+        )
+        .unwrap();
+}
+```
+
+## Including Rust sources generated by `prost`
+
+Add the following line right after the `use` section at the top of `expr/src/foo/bar/mod.rs`:
+
+```rust
+include!(concat!(env!("OUT_DIR"), "/mz_expr.foo.bar.rs"));
+```
+
+## Implementing `$T ⇔ Proto$T` mappings
+
+The implementation pattern differs between types from [classes (2) and (4)](#type-classes) and types from [class (3)](#type-classes).
```

Comment:
I prefer deleting this paragraph and just saying
"For class (2), we need to implement the `ProtoRepr` trait. Here is the implementation for `usize` as an example."
"For class(3), we need to implement the `From<&$T> for Proto$T` and `TryFrom<Proto$T> for $T` traits. As an example, here are the implementations for `MyStruct`:
"For class (4), we need to implement the `ProtoRepr` trait as well. Here is the implementation for `NaiveDate` as an example."

Reply:
Done.